File: src/test/java/info/skyblond/archivedag/controller/http/MaintainControllerTest.java
package info.skyblond.archivedag.controller.http;

import com.google.gson.Gson;
import info.skyblond.archivedag.config.EmbeddedRedisConfiguration;
import info.skyblond.archivedag.config.WebMvcConfig;
import org.hamcrest.Matchers;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;

import java.util.Map;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;

@SpringBootTest(
        classes = {EmbeddedRedisConfiguration.class, WebMvcConfig.class}
)
@ActiveProfiles("test")
@AutoConfigureMockMvc
class MaintainControllerTest {
    @Autowired
    MockMvc mockMvc;
    @Autowired
    Gson gson;

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testConfig() throws Exception {
        this.mockMvc.perform(post("/maintain/updateConfig")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(this.gson.toJson(Map.of(
                                "test-maintain-controller-key", "value"
                        ))))
                .andExpect(MockMvcResultMatchers.status().isNoContent());
        this.mockMvc.perform(get("/maintain/listConfig")
                        .param("prefix", "test-maintain-controller"))
                .andExpect(MockMvcResultMatchers.status().isOk())
                .andExpect(MockMvcResultMatchers.jsonPath("$.test-maintain-controller-key",
                        Matchers.equalTo("value")));
    }
}

File: backend/src/test/kotlin/info/skyblond/archivedag/arudaz/controller/http/MaintainControllerTest.kt
package info.skyblond.archivedag.arudaz.controller.http

import com.google.gson.Gson
import info.skyblond.archivedag.ariteg.config.EmbeddedRedisConfiguration
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

@SpringBootTest(classes = [EmbeddedRedisConfiguration::class])
@ActiveProfiles("test")
@AutoConfigureMockMvc
internal class MaintainControllerTest {
    @Autowired
    lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var gson: Gson

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testConfig() {
        mockMvc
            .perform(
                MockMvcRequestBuilders.post("/maintain/updateConfig")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(
                        gson.toJson(
                            mapOf(
                                "test-maintain-controller-key" to "value"
                            )
                        )
                    )
            )
            .andExpect(MockMvcResultMatchers.status().isNoContent)
        mockMvc
            .perform(
                MockMvcRequestBuilders.get("/maintain/listConfig")
                    .param("prefix", "test-maintain-controller")
            )
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(
                MockMvcResultMatchers.jsonPath(
                    "$.test-maintain-controller-key",
                    Matchers.equalTo("value")
                )
            )
    }
}
